File: iosApp/iosApp/AppUIAction.swift
//
//  AppUIAction.swift
//  iosApp
//
//  Created by Evgeny Tabatsky on 06.02.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import shared

class ShowToast: AppUIAction {
    let text: String
    
    init(text: String) {
        self.text = text
    }
}

class OpenSongAtVkMusic: AppUIAction {
    let music: Music
    
    init(music: Music) {
        self.music = music
    }
}

class OpenSongAtYandexMusic: AppUIAction {
    let music: Music
    
    init(music: Music) {
        self.music = music
    }
}

class OpenSongAtYoutubeMusic: AppUIAction {
    let music: Music
    
    init(music: Music) {
        self.music = music
    }
}

class SelectOrderBy: AppUIAction {
    let orderBy: OrderBy
    
    init(orderBy: OrderBy) {
        self.orderBy = orderBy
    }
}

class BackupSearchFor: AppUIAction {
    let searchFor: String
    
    init(searchFor: String) {
        self.searchFor = searchFor
    }
}


class LikeClick: AppUIAction {
    let cloudSong: CloudSong
    
    init(cloudSong: CloudSong) {
        self.cloudSong = cloudSong
    }
}

class DislikeClick: AppUIAction {
    let cloudSong: CloudSong
    
    init(cloudSong: CloudSong) {
        self.cloudSong = cloudSong
    }
}

class DownloadClick: AppUIAction {
    let cloudSong: CloudSong
    
    init(cloudSong: CloudSong) {
        self.cloudSong = cloudSong
    }
}

class UpdateDone: AppUIAction {}

File: shared/src/commonMain/kotlin/jatx/russianrocksongbook/common/state/AppUIAction.kt
package jatx.russianrocksongbook.common.state

import jatx.russianrocksongbook.common.domain.models.Music
import jatx.russianrocksongbook.common.domain.models.Warning
import jatx.russianrocksongbook.common.networking.CloudSong
import jatx.russianrocksongbook.common.networking.OrderBy

interface AppUIAction

interface KotlinUIAction: AppUIAction

data class SelectArtist(
    val artist: String,
    val callback: () -> Unit
): KotlinUIAction

object OpenSettings: KotlinUIAction

data class ReloadSettings(
    val themeVariant: ThemeVariant,
    val fontScaleVariant: FontScaleVariant
): KotlinUIAction

data class SongClick(
    val songIndex: Int
): KotlinUIAction

data class LocalScroll(
    val songIndex: Int
): KotlinUIAction

object DrawerClick: KotlinUIAction

object BackClick: KotlinUIAction

object LocalPrevClick: KotlinUIAction

object LocalNextClick: KotlinUIAction

data class FavoriteToggle(
    val emptyListCallback: () -> Unit
): KotlinUIAction

data class ConfirmDeleteToTrash(
    val emptyListCallback: () -> Unit
): KotlinUIAction

data class SaveSongText(
    val newText: String
): KotlinUIAction

object UploadCurrentToCloud: KotlinUIAction

data class SendWarning(
    val warning: Warning
): KotlinUIAction

data class CloudSearch(
    val searchFor: String,
    val orderBy: OrderBy
): KotlinUIAction

data class CloudSongClick(
    val index: Int
): KotlinUIAction

object CloudPrevClick: KotlinUIAction
object CloudNextClick: KotlinUIAction

data class SelectOrderBy(
    val orderBy: OrderBy
): KotlinUIAction

data class BackupSearchFor(
    val searchFor: String
): KotlinUIAction

data class LikeClick(
    val cloudSong: CloudSong
): KotlinUIAction

data class DislikeClick(
    val cloudSong: CloudSong
): KotlinUIAction

data class DownloadClick(
    val cloudSong: CloudSong
): KotlinUIAction

object UpdateDone: KotlinUIAction

data class OpenSongAtVkMusic(
    val music: Music
): KotlinUIAction

data class OpenSongAtYandexMusic(
    val music: Music
): KotlinUIAction

data class OpenSongAtYoutubeMusic(
    val music: Music
): KotlinUIAction

data class ShowToast(
    val text: String
): KotlinUIAction